motion/thruster_interface_asv/src/thruster_interface_asv_driver.cpp

```diff
@@ -0,0 +1,134 @@
+    double force,
+    const std::vector<double>& coeffs) {
+    return static_cast<std::uint16_t>(coeffs[0] * std::pow(force, 3) +
+                                      coeffs[1] * std::pow(force, 2) +
+                                      coeffs[2] * force + coeffs[3]);
+}
+
+void ThrusterInterfaceASVDriver::send_data_to_escs(
+    const std::vector<uint16_t>& thruster_pwm_array) {
+    constexpr std::size_t i2c_data_size =
+        1 + 4 * 2;  // 4 thrusters * (1xMSB + 1xLSB)
+    std::vector<std::uint8_t> i2c_data_array;
+    i2c_data_array.reserve(i2c_data_size);
+
+    i2c_data_array.push_back(0x00);  // Start byte
+    std::for_each(thruster_pwm_array.begin(), thruster_pwm_array.end(),
+                  [&](std::uint16_t pwm) {
+                      std::array<std::uint8_t, 2> bytes = pwm_to_i2c_data(pwm);
+                      std::copy(bytes.begin(), bytes.end(),
+                                std::back_inserter(i2c_data_array));
+                  });
+
+    // Set the I2C slave address
+    if (ioctl(bus_fd_, I2C_SLAVE, pico_i2c_address_) < 0) {
+        throw std::runtime_error("Failed to open I2C bus " +
+                                 std::to_string(i2c_bus_) + " : " +
+                                 std::string(strerror(errno)));
+        return;
+    }
+
+    // Write data to the I2C device
+    if (write(bus_fd_, i2c_data_array.data(), i2c_data_size) != i2c_data_size) {
+        throw std::runtime_error("ERROR: Failed to write to I2C device : " +
+                                 std::string(strerror(errno)));
+    }
+}
+
+std::vector<uint16_t> ThrusterInterfaceASVDriver::drive_thrusters(
+    const std::vector<double>& thruster_forces_array) {
+    // Apply thruster mapping and direction
+    std::vector<double> mapped_forces(thruster_forces_array.size());
+    for (size_t i = 0; i < thruster_parameters_.size(); ++i) {
+        mapped_forces[i] =
+            thruster_forces_array[thruster_parameters_[i].mapping] *
+            thruster_parameters_[i].direction;
+    }
+
+    // Convert forces to PWM
+    std::vector<uint16_t> thruster_pwm_array =
+        interpolate_forces_to_pwm(mapped_forces);
+
+    // Apply thruster offset and limit PWM if needed
+    for (size_t i = 0; i < thruster_pwm_array.size(); ++i) {
+        // Clamp the PWM signal
+        if (thruster_pwm_array[i] < thruster_parameters_[i].pwm_min) {
+            thruster_pwm_array[i] = thruster_parameters_[i].pwm_min;
+        } else if (thruster_pwm_array[i] > thruster_parameters_[i].pwm_max) {
+            thruster_pwm_array[i] = thruster_parameters_[i].pwm_max;
+        }
+    }
+
+    try {
+        send_data_to_escs(thruster_pwm_array);
+    } catch (const std::exception& e) {
+        std::cerr << "ERROR: Failed to send PWM values - " << e.what()
+                  << std::endl;
+    } catch (...) {
+        std::cerr << "ERROR: Failed to send PWM values - Unknown exception"
+                  << std::endl;
```

Comment:
I have mentioned this in other PRs as well, but I highly suggest looking into using spdlog for logging!